src/devices/src/virtio/block/device.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+use std::sync::Arc;
+
+use event_manager::{MutEventSubscriber, RemoteEndpoint, Result as EvmgrResult, SubscriberId};
+use kvm_ioctls::{IoEventAddress, VmFd};
+use vm_device::bus::MmioAddress;
+use vm_device::MutDeviceMmio;
+use vm_memory::GuestAddressSpace;
+use vm_virtio::devices::{VirtioConfig, VirtioMmioDevice, WithDeviceOps, WithVirtioConfig};
+use vm_virtio::Queue;
+use vmm_sys_util::eventfd::{EventFd, EFD_NONBLOCK};
+
+use crate::virtio::block::request::DiskProperties;
+use crate::virtio::block::BLOCK_DEVICE_ID;
+use crate::virtio::features::{VIRTIO_F_IN_ORDER, VIRTIO_F_RING_EVENT_IDX, VIRTIO_F_VERSION_1};
+use crate::virtio::{
+    MmioConfig, SingleFdSignalQueue, QUEUE_MAX_SIZE, VIRTIO_MMIO_QUEUE_NOTIFY_OFFSET,
+};
+
+use super::handler::QueueHandler;
+use super::simple_handler::SimpleHandler;
+use super::{build_config_space, BlockArgs, Error, Result, VIRTIO_BLK_F_FLUSH};
+
+// This Block device can only use the MMIO transport for now, but we plan to reuse large parts of
+// the functionality when we implement virtio PCI as well, for example by having a base generic
+// type, and then separate concrete instantiations for `MmioConfig` and `PciConfig`.
+pub struct Block<M: GuestAddressSpace> {
+    virtio_cfg: VirtioConfig<M>,
+    mmio_cfg: MmioConfig,
+    endpoint: RemoteEndpoint<Box<dyn MutEventSubscriber + Send>>,
+    vm_fd: Arc<VmFd>,
+    irqfd: Arc<EventFd>,
+    file_path: String,
+}
+
+impl<M: GuestAddressSpace + Clone> Block<M> {
+    pub fn new(args: BlockArgs<M>) -> Result<Self> {
+        // The queue handling logic for this device uses the buffers in order, so we enable the
+        // corresponding feature as well.
+        let device_features =
+            VIRTIO_F_VERSION_1 | VIRTIO_F_IN_ORDER | VIRTIO_F_RING_EVENT_IDX | VIRTIO_BLK_F_FLUSH;
+
+        // A block device has a single queue.
+        let queues = vec![Queue::new(args.mem, QUEUE_MAX_SIZE)];
+        let config_space = build_config_space(&args.file_path)?;
+        let virtio_cfg = VirtioConfig::new(device_features, queues, config_space);
+
+        let irqfd = EventFd::new(EFD_NONBLOCK).map_err(Error::EventFd)?;
+        args.vm_fd
+            .register_irqfd(&irqfd, args.mmio_cfg.gsi)
```

Comment:
In the spirit of #14 I'd like to see this done elsewhere (`run()`?). The *empty `Block`* can hang around inert without a `VmFd` until it *really* needs to become a live device.

Reply:
I would really like to try out this approach because it helps delegate all the setup logic to the device, as long as the necessary "resources" (such as the `VmFd` for this particular implementation) are provided by the environment.